File: tmp-mei/fgen4255027576711118260/rt/Debug.java
package rt;

import base.Void_0;

import java.util.Arrays;
import java.util.Optional;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public final class Debug implements base.Debug_0 {
  public static final Debug $self = new Debug();

  @Override public Void_0 println$imm(Object x) {
    NativeRuntime.printlnErr(toStr(x).utf8());
    return Void_0.$self;
  }

  @Override public Str identify$imm(Object x) {
    return demangle(x);
  }

  @Override public Object $hash$imm(Object x) {
    NativeRuntime.printlnErr(toStr(x).utf8());
    return x;
  }

  private static Str toStr(Object x) {
    var strMethod = java.util.Arrays.stream(x.getClass().getMethods())
      .filter(meth->
        (meth.getName().equals("str$read") || meth.getName().equals("str$readH"))
          && meth.getReturnType().equals(rt.Str.class)
          && meth.getParameterCount() == 0)
      .findAny();
    if (strMethod.isPresent()) {
      try {
        return (Str)strMethod.get().invoke(x);
      } catch(java.lang.IllegalAccessException | java.lang.reflect.InvocationTargetException err) {
        return demangle(x);
      }
    } else {
      return demangle(x);
    }
  }

  public static String demangleStackTrace(StackTraceElement[] stackTrace) {
    return Arrays.stream(stackTrace)
      .map(frame -> {
        var typeName = typeNameFromClassName(frame.getClassName()).orElse("<runtime %s>".formatted(frame.getClassName()));
        // TODO: method names etc.
        return typeName;
      })
      .collect(Collectors.joining("\n"));
  }

  private static final Pattern FEARLESS_TYPE_NAME = Pattern.compile("(.+)_(\\d+)(Impl)?(\\[.*])?$");
  private static Str demangle(Object x) {
    var strValue = x.toString();
    if (!strValue.matches(FEARLESS_TYPE_NAME.pattern())) {
      return Str.fromJavaStr(strValue);
    }

    var className = x.getClass().getName();
    var typeName = typeNameFromClassName(className)
      .orElseThrow(() -> new RuntimeException("Cannot extract type name from "+strValue));
    return Str.fromJavaStr(typeName);
  }
  private static Optional<String> typeNameFromClassName(String className) {
    var matcher = FEARLESS_TYPE_NAME.matcher(className);
    if (!matcher.matches()) {
      return Optional.empty();
    }
    var typeName = matcher.group(1);
    var nGens = Integer.parseInt(matcher.group(2), 10);
    return (typeName+"/"+nGens).describeConstable();
  }
}

File: assets/rt-js/Debug.js
import { rt$$IO } from "./IO.js";
import { rt$$Str } from "./Str.js";
import { base$$Void_0 } from "../base/Void_0.js";

export class rt$$Debug {
  static $self = new rt$$Debug();

  println$imm$1(x) {
    rt$$IO.$self.printlnErr$mut$1(rt$$Debug.toStr(x));
    return base$$Void_0.$self;
  }

  identify$imm$1(x) {
    return rt$$Debug.demangle(x);
  }

  $hash$imm$1(x) {
    rt$$IO.$self.printlnErr$mut$1(rt$$Debug.toStr(x));
    return x;
  }

  // Attempt to call x.str$read() or fallback to demangle
  static toStr(x) {
    if (x && typeof x === "object") {
      if (typeof x.str$read$0 === "function") {
        try { return x.str$read$0(); }
        catch (_) { return rt$$Debug.demangle(x); }
      }
      if (typeof x.str$readH$0 === "function") {
        try { return x.str$readH$0(); }
        catch (_) { return rt$$Debug.demangle(x); }
      }
    }
    return rt$$Debug.demangle(x);
  }

  // Demangle an object into a readable type string
  static demangle(x) {
    const strValue = x?.toString() ?? String(x);
    const match = strValue.match(rt$$Debug.FEARLESS_TYPE_NAME);
    if (!match) return rt$$Str.fromJsStr(strValue);

    const className = x?.constructor?.name ?? "<unknown>";
    const typeName = rt$$Debug.typeNameFromClassName(className);
    if (!typeName) throw new Error("Cannot extract type name from " + strValue);
    return rt$$Str.fromJsStr(typeName);
  }

  // Convert className like Foo_1Impl to Foo/1
  static typeNameFromClassName(className) {
    const match = className.match(rt$$Debug.FEARLESS_TYPE_NAME);
    if (!match) return null;
    const typeName = match[1];
    const nGens = parseInt(match[2], 10);
    return `${typeName}/${nGens}`;
  }

  // Optional: demangle a stack trace
  static demangleStackTrace(stackTrace) {
    return (stackTrace || []).map(frame => {
      return rt$$Debug.typeNameFromClassName(frame?.className) ?? `<runtime ${frame?.className}>`;
    }).join("\n");
  }
}

// Regex for Fearless type mangling: Foo_1Impl, Bar_2[], etc.
rt$$Debug.FEARLESS_TYPE_NAME = /^(.+)_(\d+)(Impl)?(\[.*])?$/;
